tools/update.php

```diff
@@ -0,0 +1,747 @@
+        }
+    }
+    if (empty($release_dir)) {
+        die("ERROR: Could not find downloaded files in $dst_dir" . PHP_EOL);
+    }
+    // Use rsync to overwrite files in $loris_root
+    echo '[*] Overwriting old source code files.'  . PHP_EOL;
+    $cmd = 'rsync -r ' . escapeshellarg($release_dir) . ' ' .
+        escapeshellarg($loris_root_dir);
+    return doExec($cmd);
+}
+
+/**
+ * Ensures required packages and their repositories are known to the server
+ * and are up-to-date.
+ *
+ * @param array $requirements Apt packages required by LORIS.
+ *
+ * @return bool True if all packages installed and up-to-date. False otherwise.
+ */
+function updateRequiredPackages($requirements) : bool
+{
+    // we need 3rd party PPA to get the latest PHP on Ubuntu
+    echo '[*] Adding external PPAs...' . PHP_EOL;
+    // -y flag required to suppress a message from the author
+    // TODO: check if these already exist before adding
+    exec('sudo apt-add-repository ppa:ondrej/php -y');
+    exec('sudo apt-add-repository ppa:ondrej/apache2 -y');
+
+    echo '[*] Updating apt package list...' . PHP_EOL;
+    exec('sudo apt-get update');
+    // die unless all required packages are installed and up-to-date
+    if (!(installMissingRequirements($requirements))
+        && (installAptPackages($requirements, $upgrade_mode = true))
+    ) {
+        die(
+            'Could not upgrade all required packages. Exiting.'
+            . PHP_EOL
+        );
+    }
+    return true;
+}
+
+/**
+ * Returns a list of SQL patch files to apply based on LORIS version
+ * differences.
+ *
+ * @param string $loris_root   The directory where LORIS is installed
+ * @param string $version_from Pre-update LORIS version
+ * @param string $version_to   Post-update LORIS version
+ *
+ * @return array Patch files to be applied
+ */
+function getPatchesFromVersion($loris_root, $version_from, $version_to) : array
+{
+    // Semantic versioning: 0 = major, 1 = minor, 2 = bugfix
+    define('MAJOR', 0);
+    define('MINOR', 1);
+    define('BUGFIX', 2);
+    // Convert string representation of version into arrays of ints
+    // TODO: check for dev branches
+    $from_versions = array_map('intval', explode('.', $version_from));
+    $to_versions   = array_map('intval', explode('.', $version_to));
+    // Calculate difference between old version and latest
+    $diff_major = $to_versions[MAJOR] - $from_versions[MAJOR];
+    $diff_minor = $to_versions[MINOR] - $from_versions[MINOR];
+    if ($diff_major < 0) {
+        echo '[!] Your version of LORIS is ahead of the latest release. If you'
+        . ' are not a developer working on the bleeding edge of LORIS, then '
+        . 'something has gone very wrong. No SQL patches will be displayed.'
+        . PHP_EOL;
+        return array();
+    }
+    echo "[**] Latest version $version_to is ahead of installed $version_from "
+        . "by $diff_major MAJOR release(s), $diff_minor MINOR release(s)."
+        . PHP_EOL;
```

Comment:
This could end up as `Latest version 22.0 is ahead of installed 21.3 by 1 MAJOR release(s), -3 MINOR release(s).`

This negative minor can be quite confusing. I suggest to remove the minor part when major is > 0.